### DISCUSSION TITLE
How to use  app.storage or any other ways to share data between  UI.tabs per session

### DISCUSSION THREAD

**Discussion by bajirama** [zauberzeug/nicegui#3366] [Q&A]
### Question

I hope this message finds you well. I am reaching out to seek guidance on implementing a feature within our web application. Our goal is to enable data entered in one tab to be accessible across multiple tabs, confined to a single user session. Although I have attempted to utilize `app.storage`, my efforts have not yielded the desired outcome.

To provide a clearer picture, I have attached the sample code below that illustrates the current setup. I believe the solution may involve leveraging `localStorage` and event listeners to synchronize session data across tabs, as suggested in a method I found. However, I am uncertain about the best way to integrate this into our existing framework.

Any assistance or direction you could provide on this matter would be greatly appreciated. I am eager to learn and apply the most effective approach to enhance our application's functionality.

Thank you for your time and consideration.
```py
from nicegui import app, ui
def main_page():
  # Function to get custom variable from storage
  def get_custom_var():
    svar =  app.storage.general.get("awsdetails")
    print(svar)
    return svar

  # Function to set custom variable in storage
  def set_custom_var(aValue):
    app.storage.general.update({'awsdetails': aValue})
    print("Saved")


  async def update_custom_var():
    aws_detail = [{ "env_type": env_type.value,
                   "acc_no": acc_no.value,
                   "acc_name" : acc_name.value,
                   "aws_rgn" : aws_rgn.value
                 }]
    set_custom_var(aws_detail)

  def on_tab_change(event):
      print(f"Selected tab: {event.value}")
      # Add your custom logic here
  def tab2view():
      accdetail = get_custom_var()
      print("In Tab2")
      print(accdetail)
      accUser = []
      accColDef = [
          {'headerName': "Account Type", "field": "Acctype", 'checkboxSelection': True},
          {'headerName': "Account Number", "field": "AccNo"},
          {'headerName': "Account Name", "field": "AccName"},
        ]
      with ui.row().classes('w-full'):
              awsDetailsTable = ui.aggrid({
                  'defaultColDef': {'flex': 1},
                  'columnDefs': accColDef,
                  'rowData': accdetail,
                  'rowSelection': 'single'
              }).classes('w-full')
      tab2_input = ui.input('Enter something in Tab 2')
          # tab2_output = ui.label(f'Tab 2 input: {app.storage.general["tab2_input"]}')

  def tab3View():
      accdetail = get_custom_var()
      accColDef = [
          {'headerName': "Account Type", "field": "Acctype", 'checkboxSelection': True},
          {'headerName': "Account Number", "field": "AccNo"},
          {'headerName': "Account Name", "field": "AccName"},
        ]
      with ui.row().classes('w-full'):
              accDetailsTable = ui.aggrid({
                  'defaultColDef': {'flex': 1},
                  'columnDefs': accColDef,
                  'rowData': accdetail,
                  'rowSelection': 'single'
              }).classes('w-full')
      tab3_input = ui.input('Enter something in Tab 3')


# Create a NiceGUI page to interact with the custom variable
  with ui.left_drawer().style('background-color: #F8F8F8').props("elevated v-model='leftDrawerOpen'"):
      with ui.tabs().props('vertical').classes('w-full').on('change', on_tab_change)as tabs:
           one = ui.tab('One')
           two = ui.tab('Two')
           three = ui.tab('Threee')
  with ui.tab_panels(tabs, value=one).classes('w-full'):
      with ui.tab_panel(one):
          with ui.column().classes('items-center'):
              with ui.card().classes('centered'):
                  ui.label('Custom Variable Storage Example')

                  # Input field to update the custom variable
                  env_type = ui.input(label='Acc type')
                  acc_no = ui.input(label='Acc_No')
                  acc_name = ui.input(label='Acc Name')
                  aws_rgn = ui.input(label='Region')
                  # Button to set the custom variable
                  ui.button('Set Customer Variable', on_click=update_custom_var)

      with ui.tab_panel(two):
            tab2view()
      with ui.tab_panel(three):
            tab3View()

# Create the main page
ui.page("/")
main_page()
# Start the NiceGUI application
ui.run(port=9090)
```

**Comment by rodja** [comment]
Have you tried using `app.storage.user`? It is exactly build for sharing data between different tabs of the same session. If that does not solve your problem on the spot, I would kindly ask you to create a minimum reproducible code example. This would allow us and the community to help more efficiently. Thanks!
[Why is it important to provide a minimal reproducible example?](https://github.com/zauberzeug/nicegui/wiki/FAQs#why-is-it-important-to-provide-a-minimal-reproducible-example)

**Comment by bajirama** [comment]
I am currently working on a feature for our application that involves the app.storage.user session. The first part of the feature, which collects and stores key-value pairs in one tab, is functioning correctly. However, I am seeking assistance with the second tab.
The goal for the second tab is to retrieve all keys from the app.storage, allow the user to select a key, and then display the corresponding value in a textarea. Additionally, there should be an 'Update' button that updates the value from the textarea to the app.storage. In the future, I plan to implement a 'Delete' button to remove the selected key's value.
I would greatly appreciate any guidance or code snippets that could help me achieve this functionality. Your expertise has always been invaluable to our projects,

**Need Help with** 
1.	How to update the  app.storage.user per key  ?
2.	How to delete the app.storage.user per key per row ?
3.	How to update /refresh the tab values on the tab change event  ?

**Sample Data  -app.storage.user** 

```py
"user": [{"username":"User1","authenticated":true}],
"accdet": [{'acc_type': 'Test1', 'acc_no': '643738098711', 'AccName': 'Test Account1', 'region': 'US Region'},
           {'acc_type': 'Test2', 'acc_no': '643738098711', 'AccName': 'Test Account2', 'region': 'US Region'},
           {'acc_type': 'Test3', 'acc_no': '643738098711', 'AccName': 'Test Account3', 'region': 'US Region'}

```
**Minimum reproducible code example**

```py
from nicegui import ui, app

@ui.page('/')

def storagetest() :
   def save2Session():
       sKey = keyvalue.value
       sValue = Valvalue.value

       sValue = {sKey:sValue}
       print(f"svalue-{sValue}")
       print(f"skey-{sKey}")


       if sValue :
            if app.storage.user.get(sKey):
               app.storage.user[sKey].clear()

            app.storage.user.update(sValue)
            app.storage.tab.update(sValue)
            app.storage.client.update(sValue)
            print("Upated")
       else:
           print("No Value Updated")

   def getKeyvalue(event):
       if event.value  :
           sKeyval = app.storage.user.get(event.value)
           Valvalue2.set_value( sKeyval)
       else:
           Valvalue2.set_value("No Selection")

   def updatestroage():
       print("Need help in updating the select key")


   with ui.tabs() as tabs:
        tab1 = ui.tab('Tab 1')
        tab2 = ui.tab('Tab 2')

   with ui.tab_panels(tabs, value=tab1):
      with ui.tab_panel(tab1):
        with ui.row().classes("w-full")  :
           label1 = ui.label('Shared value1: ')
           keyvalue  = ui.input('Enter Key').classes("pl-6 w-5/6")
           Valvalue  = ui.textarea('Enter Value').classes("pl-6 w-5/6")
           ui.button('Add', on_click=save2Session, icon='add').classes('m-2')

      with ui.tab_panel(tab2):
          label2 = ui.label('Shared value2: ')
          if app.storage.user :
             asessionkey = app.storage.user.key()
          else :
              asessionkey = []
          sessionkey  = ui.select(options=asessionkey, value="", label='User Key',on_change=getKeyvalue).classes('pl-6 w-5/6')
          Valvalue2   = ui.textarea('Enter Value').classes("pl-6 w-5/6")
          ui.button('Update', on_click=updatestroage, icon='update').classes('m-2')

ui.run(storage_secret='MY_PORTAL',port=9090)
```

**Reply by falkoschindler** [reply]
Hi @bajirama,

When started, your example raises a NameError:
```
Traceback (most recent call last):
  File "/Users/falko/Projects/nicegui/nicegui/events.py", line 413, in handle_event
    result = handler(arguments) if expects_arguments else handler()
             ^^^^^^^^^^^^^^^^^^
  File "/Users/falko/Projects/nicegui/test.py", line 31, in getKeyvalue
    Valvalue2.set_value("No Selection")
    ^^^^^^^^^
NameError: cannot access free variable 'Valvalue2' where it is not associated with a value in enclosing scope
```

Apart from that: Are the tabs relevant to your problem? If not, let's remove them.

**Comment by falkoschindler** [comment]
Here's a simple example for such a user storage editor:
```py
@ui.page('/')
def page():
    with ui.tabs() as tabs:
        tab1 = ui.tab('Add Entry')
        tab2 = ui.tab('Update Entries')

    with ui.tab_panels(tabs, value=tab1):
        with ui.tab_panel(tab1):
            def add():
                app.storage.user[new_key.value] = new_value.value
                new_key.value = ''
                new_value.value = ''
                select.set_options(list(app.storage.user))

            new_key = ui.input('Key')
            new_value = ui.textarea('Value')
            ui.button('Add', on_click=add)

        with ui.tab_panel(tab2):
            select = ui.select(list(app.storage.user), on_change=lambda e: value.set_value(app.storage.user[e.value]))
            value = ui.textarea('Value')
            ui.button('Update', on_click=lambda: app.storage.user.update({select.value: value.value}))
```

**Comment by bajirama** [comment]
Thank you for your update. It works well, but I plan to use this feature extensively. Specifically, I intend to manage at least 5 to 6 items, each of which might need updating at different times. Therefore, I need to understand how to clear or update a single item in app.storage.user without updating the entire JSON file.

For example, App.storage might contain the following structure, with different workflows updating individual items within app.storage.user session data. In your current example, the whole JSON file is updated. Is there a way to update or clear just one item in app.storage.user?


```
"user": [{"username":"User1","authenticated":true}],
"accdet": [{'acc_type': 'Test1', 'acc_no': '643738098711', 'AccName': 'Test Account1', 'region': 'US Region'},
           {'acc_type': 'Test2', 'acc_no': '643738098711', 'AccName': 'Test Account2', 'region': 'US Region'},
           {'acc_type': 'Test3', 'acc_no': '643738098711', 'AccName': 'Test Account3', 'region': 'US Region'}
"item2":  [{'items2_Key1_row1': 'Item2_val2_row1' , 'items2_Key2_row1': 'Item2_val2_row1','items2_Key3_row1': 'Item2_val3'}],
```

**Reply by falkoschindler** [reply]
How (and why) would you _not_ update the whole JSON file? Do you want to reduce the amount of write access to the disk storage? Or do you fear race conditions?

**Reply by bajirama** [reply]
I want to reduce the write access  to disk storage.   Also I  found some useful functions to do that like app.storage.user.popitem() and  it is  working

**Reply by falkoschindler** [reply]
Reducing disk access for NiceGUI's storage module sounds like a new feature request. Let's open a new thread focusing on this specific subject. I still think that you can't _partly_ write JSON files. But we could throttle the file access.

**Comment by bajirama** [comment]
I'm in need of assistance with this issue and find myself going down a rabbit hole, unable to find a solution. Typically, with NiceGUI, a few hours of work yields some resolution, but I've been struggling for days now.
 Here's the scenario: I have two tabs on the left. In the first tab, I enter a value that is displayed in the AGGrid. The key value will be the screen name, which remains constant. The **'Add'** button will append the value to the grid, while the '**Save'** button will store the value in app.storage.user.
 
In the second tab, I need to reference the same data entered in the first tab. Any changes in Tab 1 should be reflected in Tab 2. Additionally, I need the value from Tab 1 to be reflected in other tabs on the left panel.
 For clearer code, I plan to separate each tab into its own module, like Tab 2. Currently, I'm using app.storage.user to save values in the **.nicegui** folder. The values are saved, and changes are updated as needed in app.storage.user. 

However, my challenge is dynamically refreshing and retrieving the latest values from app.storage.user. 
Once resolved, I also plan to add a dropdown at the top section of the left panel, allowing selection of any value to reset all tabs for new criteria. 
Could you advise on how to achieve dynamic refresh for tabs? I've attached screenshots, minimal viable code, and sample values from app.storage.user for reference.


**First time entry**  
**Tab 1  - Add Entry**
 
![image](https://github.com/user-attachments/assets/610cb62b-f6f5-4f9c-8dc2-d439c317e2b9)


**Tab2 – Update Entry** 
 
![image](https://github.com/user-attachments/assets/6fbcc59e-cc86-4a38-b30a-832d5b58f9be)


**After Screen Refresh**  
**Tab 1  - Add Entry**
 
![image](https://github.com/user-attachments/assets/9db4680a-ec15-4ef6-be3b-68a1943d9486)


**Tab2 – Update Entry** 
 
![image](https://github.com/user-attachments/assets/92b9a8b8-24f9-49ca-8f55-8f7c55a41d9f)

**Sample Date from app.storage.user** 

```json
{"MyKeyName": [
        {
            "acc_type": "Value11",
            "acc_no": "Value12",
            "AccName": "Value13",
            "region": "Value14"
        },
        {
            "acc_type": "Value21",
            "acc_no": "Value22",
            "AccName": "Value23",
            "region": "Value24"
        },
        {
            "acc_type": "Value31",
            "acc_no": "Value32",
            "AccName": "Value33",
            "region": "Value34"
        },
        {
            "acc_type": "Value41",
            "acc_no": "Value42",
            "AccName": "Value43",
            "region": "Value44"
        }
    ]
}
```
**Main code - refreshtest.py**

```py
from nicegui import ui, app ,Client
import _tabfun

@ui.page('/')
def page(client: Client):
    accdetails = []
    @ui.refreshable
    def refreshStoage() -> None:
        accdetails = app.storage.user.get("MyKeyName")
        print(f"refresh {accdetails}")
        ui.label("Upated")

    def savestorage():
        app.storage.user[new_key.value] = accdetails
        refreshStoage.refresh()

    with ui.left_drawer().style('background-color: #F8F8F8').props("elevated v-model='leftDrawerOpen'"):

        with ui.tabs().props('vertical').classes('w-full') as tabs:
            tab1 = ui.tab('Add Entry')
            tab2 = ui.tab('Update Entries')
    with client.content:
      with ui.tab_panels(tabs, value=tab1):
          with ui.tab_panel(tab1):

              def add():
                  laccdetails = [{'acc_type': new_value1.value,
                                 'acc_no': new_value2.value,
                                 'AccName': new_value3.value,
                                 'region': new_value4.value,
                                 }]
                  accdetails.extend(laccdetails )
                  AccDetTable.options['rowData'] = accdetails
                  AccDetTable.update()


              with ui.card().style('width: 1400px; max-width: none'):
                 with ui.row().classes("w-full"):
                     accDetColDef = [
                            {'headerName': "Acc Type ", "field": "acc_type", 'checkboxSelection': True},
                            {'headerName': "Account Number", "field": "acc_no"},
                            {'headerName': "Account Name", "field": "AccName"},
                            {'headerName': "Region", "field": "region"},
                        ]
                     AccDetTable = ui.aggrid({
                            'defaultColDef': {'flex': 1},
                            'columnDefs': accDetColDef,
                            'rowData': accdetails,
                           'rowSelection': 'single',
                         }).classes('max-h-40')

                 with ui.row().classes("w-full"):
                    new_key    = ui.input("KeyValue", value= "MyKeyName",).props("disable")
                    new_value1 = ui.input('Value1')
                    new_value2 = ui.input('Value2')
                    new_value3 = ui.input('Value3')
                    new_value4 = ui.input('Value4')
                    refreshStoage()
                    ui.button('Add', on_click=add)
                    ui.button("Save", on_click=savestorage)

          with ui.tab_panel(tab2):
            _tabfun.tab2scr()


ui.run(storage_secret='MY_PORTAL')
```

**second tab function - _tabfun.py** 
```py
from nicegui import ui, app

def tab2scr():
    laccDet = app.storage.user.get("MyKeyName")
    with ui.row().classes('w-full'):
      with ui.card().style('width: 1400px; max-width: none'):
         ui.notify("Test")
         ui.label(laccDet)
         accDetColDef = [
                {'headerName': "Acc Type ", "field": "acc_type", 'checkboxSelection': True},
                {'headerName': "Account Number", "field": "acc_no"},
                {'headerName': "Account Name", "field": "AccName"},
                {'headerName': "Region", "field": "region"},
              ]
         AccDetTable = ui.aggrid({
           'defaultColDef': {'flex': 1},
           'columnDefs': accDetColDef,
           'rowData': laccDet,
           'rowSelection': 'single',
        }).classes('max-h-40')
```

**Reply by rodja** [reply]
@bajirama this looks like a different kind of problem than the original question. But I can't say for sure. It's just too long and hard to understand. If it is different, please start a new question. In either case, try to create a minimum reproducible code example. This would allow us and the community to help more efficiently. Thanks!
[Why is it important to provide a minimal reproducible example?](https://github.com/zauberzeug/nicegui/wiki/FAQs#why-is-it-important-to-provide-a-minimal-reproducible-example)

**Comment by bajirama** [comment]
It is same issue I just created a new thread as you requested. (https://github.com/zauberzeug/nicegui/discussions/3414)
also I have the minimum reproducible code example. in  thread  also pasted the screen shot of how it looks
